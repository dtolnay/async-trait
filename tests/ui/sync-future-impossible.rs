use async_trait::async_trait;

#[async_trait]
trait Trait {
    #[future_is[Sync]]
    async fn method(&mut self);
}

struct Struct<T> {
    value: T,
}

#[async_trait]
impl<T: std::fmt::Display + Send> Trait for Struct<T> {
    #[future_is[Sync]]
    async fn method(&mut self) {
        println!("{}", self.value);
    }
}

fn main() {}
