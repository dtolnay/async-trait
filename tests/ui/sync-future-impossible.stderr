error: future cannot be shared between threads safely
  --> $DIR/sync-future-impossible.rs:16:32
   |
16 |       async fn method(&mut self) {
   |  ________________________________^
17 | |         println!("{}", self.value);
18 | |     }
   | |_____^ future created by async block is not `Sync`
   |
note: captured value is not `Sync`
  --> $DIR/sync-future-impossible.rs:17:24
   |
17 |         println!("{}", self.value);
   |                        ^^^^ has type `&mut Struct<T>` which is not `Sync`
   = note: required for the cast to the object type `dyn std::future::Future<Output = ()> + std::marker::Send + std::marker::Sync`
help: consider further restricting this bound
   |
14 | impl<T: std::fmt::Display + Send + std::marker::Sync> Trait for Struct<T> {
   |                                  ^^^^^^^^^^^^^^^^^^^
